ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryReplicaSyncUpService.java

```diff
@@ -0,0 +1,197 @@
+  public void waitBootstrapCompleted(String partitionName) throws InterruptedException {
+    CountDownLatch latch = partitionToBootstrapLatch.get(partitionName);
+    if (latch == null) {
+      logger.info("Skipping bootstrap for existing partition {}", partitionName);
+    } else {
+      logger.info("Waiting for new partition {} to complete bootstrap", partitionName);
+      latch.await();
+      partitionToBootstrapLatch.remove(partitionName);
+      if (!partitionToBootstrapSuccess.remove(partitionName)) {
+        throw new IllegalStateException("Partition " + partitionName + " failed on bootstrap.");
+      }
+      logger.info("Bootstrap is complete on partition {}", partitionName);
+    }
+  }
+
+  @Override
+  public boolean updateLagBetweenReplicas(ReplicaId chaser, ReplicaId precursor, long lagInBytes) {
+    boolean updated = false;
+    if (replicaToLagInfos.containsKey(chaser)) {
+      replicaToLagInfos.get(chaser).updateLagInfo(precursor, lagInBytes);
+      logger.debug(replicaToLagInfos.get(chaser).toString());
+      updated = true;
+    }
+    return updated;
+  }
+
+  @Override
+  public boolean isSyncUpComplete(ReplicaId replicaId) {
+    RemoteReplicaLagInfos lagInfos = replicaToLagInfos.get(replicaId);
+    if (lagInfos == null) {
+      throw new IllegalStateException(
+          "Replica " + replicaId.getPartitionId().toPathString() + " is not found in AmbryReplicaSyncUpService!");
+    }
+    return lagInfos.hasSyncedUpWithEnoughPeers();
+  }
+
+  @Override
+  public void onBootstrapComplete(String partitionName) {
+    partitionToBootstrapSuccess.put(partitionName, true);
+    countDownLatch(partitionName);
+  }
+
+  @Override
+  public void onBootstrapError(String partitionName) {
+    countDownLatch(partitionName);
+  }
+
+  /**
+   * clean up in-mem maps
+   */
+  void reset() {
+    partitionToBootstrapLatch.clear();
+    partitionToBootstrapSuccess.clear();
+    replicaToLagInfos.clear();
+  }
+
+  /**
+   * Count down the latch associated with given partition
+   * @param partitionName the partition whose corresponding latch needs to count down.
+   */
+  private void countDownLatch(String partitionName) {
+    CountDownLatch latch = partitionToBootstrapLatch.get(partitionName);
+    if (latch == null) {
+      throw new IllegalStateException("No bootstrap latch is found for partition " + partitionName);
+    } else {
+      latch.countDown();
+    }
+  }
+
+  private class RemoteReplicaLagInfos {
+    // keep lag map here for tracking progress
+    private final ConcurrentHashMap<ReplicaId, Long> localDcPeerReplicaAndLag = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<ReplicaId, Long> remoteDcPeerReplicaAndLag = new ConcurrentHashMap<>();
+    // refer to those replicas that current replica has caught up with
+    private final Set<ReplicaId> localDcCaughtUpReplicas = ConcurrentHashMap.newKeySet();
+    private final Set<ReplicaId> remoteDcCaughtUpReplicas = ConcurrentHashMap.newKeySet();
```

Comment:
If the catchup threshold is a total across local and remote replicas, is there any benefit to tracking them separately?  Or conversely, would it make sense to split the target into local and remote targets and require both to be true?

Reply:
Usually, for new replica, catching up with all local peer replicas should suffice. In case that some local replicas are down (especially when it occurs in DC that only has one peer replica), new replica then is supposed to catch up with enough peers in remote DCs. We didn't separate local DC and remote DC because it usually catches up with local DC replicas first.  As mentioned before, we add remote replicas here in case local replicas are unavailable.

Reply:
That makes sense.  But then separating the caught-up-replica set into local and remote doesn't seem to be add much.

Reply:
You are right, we technically can combine them together. For now, I track them separately for debugging purpose. Since there is not too much overheard by doing so, let's temporarily keep them here and simplify them if needed in the future.